include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,109 @@
+#ifndef KOKKOS_UTILS_CALLBACKS_TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS_TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+template <typename TimerType>
+concept Startable = requires (TimerType& timer) {
+    { timer.start() } -> std::same_as<void>;
+};
+
+template <typename TimerType, typename EventType>
+concept StartableWithEvent = requires (TimerType& timer, const EventType& event) {
+    { timer.start(event) } -> std::same_as<void>;
+};
+
+template <typename TimerType>
+concept Stoppable = requires (TimerType& timer) {
+    { timer.stop() } -> std::same_as<void>;
+};
+
+template <typename TimerType, typename EventType>
+concept StoppableWithEvent = requires (TimerType& timer, const EventType& event) {
+    { timer.stop(event) } -> std::same_as<void>;
+};
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The begin event is selected by requiring it to be a match for @p BeginEndMatcherType
+ * while no begin event has been matched before. The end event is selected by requiring
+ * it to be a match for @p BeginEndMatcherType after a begin event has been matched.
+ *
+ * The listener can be reset by using the member function @ref reset. After calling
+ * @ref reset, a new begin event can be matched.
+ */
+template <
+    Matcher BeginEndMatcherType,
+    typename TimerType
+>
+class TimerListener
+{
+public:
+    //! @todo Check if this constructor is needed.
```

Comment:
Have you checked this ?

Reply:
I've just tried to check it. It seems we currently need it, and it seems the issue is that
- the `matcher` member is private
- the order of the members is such that `matched_begin` and `matched_begin` "are in the middle"

If I make all the members public and begin with `matcher` and `timer`, the constructor can be deleted.

It seems the issue of private members comes up in other classes as well. So probably best discussed live if we want to make all the members public.